enhancements/ingress/tunable-router-buffer-sizes.md

```diff
@@ -0,0 +1,165 @@
+the size of per-session HAProxy HTTP message buffers. The [OpenShift Router](https://github.com/openshift/rotuer)
+sets this value to `32768` bytes by default.
+
+The [HAProxy tune.maxrewrite](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#tune.maxrewrite) setting specifies
+how much memory HAProxy will reserve from the buffer (with length `tune.bufsize`) for HTTP header rewriting and appending.
+The [OpenShift Router](https://github.com/openshift/rotuer) sets this value to `8192` bytes by default.
+
+The IngressController API is extended by adding an optional `HeaderBuffer` field with
+type `*IngressControllerHeaderBuffer` to `IngressControllerSpec`:
+
+```go
+type IngressControllerSpec struct {
+	// ...
+
+	// HeaderBuffer defines parameters for buffer size values. If this
+	// field is empty, the default values set within the IngressController
+	// will be used.
+	//
+	// +optional
+	HeaderBuffer *IngressControllerHeaderBuffer `json:"headerBuffer,omitempty"`
+}
+```
+
+The `IngressControllerHeaderBuffer` type has 2 `int` fields, one for
+`HeaderBufferSize` and one for `HeaderBufferMaxRewrite`.
+
+```go
+type IngressControllerHeaderBuffer struct {
+	// buffersize describes how much memory should be reserved
+	// (in bytes) for IngressController connection sessions.
+	//
+	// +kubebuilder:validation:required 
+	// +kubebuilder:validation:minimum=<tbd>
+	// +required
+	HeaderBufferSize int `json:"headerBufferSize,omitempty"`
+
+	// buffermaxrewrite describes how much memory should be reserved
+	// (in bytes) from HeaderBufferSize for HTTP header rewriting
+	// and appending.
+	//
+	// +kubebuilder:validation:required
+	// +kubebuilder:validation:minimum=<tbd>
+	// +required
+	HeaderBufferMaxRewrite int `json:"headerBufferMaxRewrite,omitempty"`
+}
+```
+
+Proper environment variables will be added to the Router deployment to allow for the `HeaderBufferSize` and `HeaderBufferMaxRewrite` values to be
+populated in the HAProxy configuration template.
+
+The [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) will need to be modified to handle both the API changes
+and setting the new Router environment variables.
+
+Integration tests will also need to be written to ensure that the new IngressController API fields work as intended (and that the router
+respects buffer limits without degrading).
+
+### User Stories
+
+#### As a Cluster Administrator, I want to edit the HAProxy configuration by setting "tune.maxrewrite" and "tune.bufsize" so that the Router can accommodate requests with large header sizes
+
+The necessary Router and API changes will be made to allow cluster administrators to tweak these settings on a per-IngressController level.
+
+### Implementation Details/Notes/Constraints
+
+Header buffer size values should be validated to make sure they are at least sane. For example, `HeaderBufferSize` values of `0` or `MAXINT` should most likely
+be rejected.
```

Comment:
In particular, `headerBufferSize - headerBufferMaxRewrite` must be reasonable.  Unfortunately we cannot validate that in the API.  However, the operator could validate the values and set an appropriate status condition if the ingresscontroller specified invalid values.  

One implication of having validation in the operator is that this validation can special-case the default ingresscontroller.  That is, the operator could reject values that could break OAuth for the default ingresscontroller, but allow those values for other ingresscontrollers.

Reply:
I mentioned these thoughts in the enhancement. It would make sense to me to have the ingress operator prevent a cluster admin from breaking Oauth on the default ingress controller with this new API field.